lib/clamp-device.in

```diff
@@ -190,11 +169,31 @@ if [ $KMDUMPLLVM == "1" ]; then
   cp $2.linked.bc ${KMDUMPDIR}/dump.linked.bc
 fi
 
+# Invoke HCC-specific opt passes
+if [ "$HCC_TILECHECK" == "ON" ]; then
+    $OPT -load $LIB/LLVMSelectAcceleratorCode@CMAKE_SHARED_LIBRARY_SUFFIX@ \
+        -select-accelerator-code -dce -globaldce < $2.linked.bc -o $2.selected.bc
+    $OPT -always-inline -tile-uniform < $2.selected.bc -o $2.selected.bc
+else
+    $OPT -load $LIB/LLVMSelectAcceleratorCode@CMAKE_SHARED_LIBRARY_SUFFIX@ \
+        -select-accelerator-code -dce -globaldce < $2.linked.bc -o $2.selected.bc
+    $OPT -always-inline < $2.selected.bc -o $2.selected.bc
+fi
+
+# error handling for HCC-specific opt passes
+RETVAL=$?
+if [ $RETVAL != 0 ]; then
+  echo "Generating AMD GCN kernel failed in HCC-specific opt passes for target: $AMDGPU_TARGET"
+  exit $RETVAL
+fi
+
+if [ $KMDUMPLLVM == "1" ]; then
+  cp $2.selected.bc ${KMDUMPDIR}/dump.selected.bc
+fi
+
 # Optimization notes:
 #  -disable-simplify-libcalls:  prevents transforming loops into library calls such as memset, memcopy on GPU
-# Teach LLVM inliner we do want all functions used by a kernel be inlined
-# by assigning a big iniling threshold value
-$OPT -inline -inline-threshold=1048576 -mtriple amdgcn--amdhsa-amdgiz -mcpu=$AMDGPU_TARGET -infer-address-spaces -amdgpu-internalize-symbols -disable-simplify-libcalls $KMOPTOPT -verify $2.linked.bc -o $2.opt.bc
+$OPT -mtriple amdgcn--amdhsa-amdgiz -mcpu=$AMDGPU_TARGET -infer-address-spaces -amdgpu-internalize-symbols -disable-simplify-libcalls $KMOPTOPT -verify $2.selected.bc -o $2.opt.bc
```

Comment:
@AlexVlx if i understand you correctly, `-always-inline` in the previous `opt` invocation is supposed to replace `-inline -inline-threshold=1048576` here, right? but i think `-always-inline` will only inline functions marked with `alwaysinline` attribute, and will *NOT* inline functions without that attribute. but after linking with `ROCDL` we need all kernels be fully inlined as of now, so I think `-inline -inline-threshold` still need to be present?